plugins/tradingAdvisor/tradingAdvisor.js

```diff
@@ -86,6 +86,9 @@ Actor.prototype.setupStrategy = function() {
 // HANDLERS
 // process the 1m candles
 Actor.prototype.processCandle = function(candle, done) {
+  // notify strategy about one min
+  this.strategy.updateOneMin(candle);
```

Comment:
Though maybe do this bellow line 92? The docs do state that this.candle reflects the current candle. In your proposal it reflects the old one (since you run a strat function before setting it). Also we should probably document/think about how we want the order of functions to be called.

If a strategy has async indicators Gekko should defer all other 1min candles until after they are calculated, but the buffering of additional 1min candles while the async indicators are calculated happens below this point. Meaning that (depending on how the node stream write candles) this might trigger `updateOneMin` functions of candles of candles that happen after a bigger candle might arrive first.

For example: If you you have a strategy that:

- uses a tulip indicator
- runs over 10min candles
- has a  `updateOneMin` function
- is running in a backtest (or realtime markets with burst of candles, see #2461)
- Candles are from `:00`, `:10`, `:20`, etc.

I am not sure if right now you are guarenteed to get your 10min candle on 10:10 before any 1 min candles of 10:11 fire (since they might buffer below this code). Whether this matters depends on your strategy.

Reply:
I am not 100% sure of the above, Gekko uses [Node streams](https://nodejs.org/api/stream.html) to pump candles into plugins (they end up in  plugins/tradingAdvisor/tradingAdvisor.js:88). And with a stream you pass a callback once you are done with consuming (a candle), the callback is only called after Gekko [is done with calculating all the async stuff](https://github.com/askmike/gekko/blob/b99c5a2f9ad766b203608b006f99a9f7e22713f0/plugins/tradingAdvisor/tradingAdvisor.js#L105). However the streams API has changed and I am currently not sure how much buffering is candle guarenteed to happen before you call this callback function. As such Gekko implements another layer of buffering here: https://github.com/askmike/gekko/blob/develop/plugins/tradingAdvisor/asyncIndicatorRunner.js#L28-L36

Reply:
As I understood the code we have a flow: 
market stream -> 1min candle -> batcher buffer -> emit big candle -> cal. indicators sync/async -> strategy.update -> done() get new 1 min candle from the stream.

than we can put updateOneMin just before the buffer
market stream -> 1min candle -> strategy.updateOneMin -> batcher buffer -> emit big candle -> cal. indicators sync/async -> strategy.update -> done() get new candle from the stream.

The strategy gets always first 1 min candle and than we get the buffered big candle. And when buffering takes very long, than the programm still waits for the done() and doesn't get new candle. 

As long the updateOneMin has sync code there should be no problem. If you start async code in  updateOneMin with callbacks than yes we might get into trouble. But now the update-method also expects only sync code. 

And yes putting   this.strategy.updateOneMin(candle); after line 92 is good.

Reply:
> The strategy gets always first 1 min candle and than we get the buffered big candle. And when buffering takes very long, than the programm still waits for the done() and doesn't get new candle.

To make sure we are on the same line, I am not talking about any buffering from the candleBatcher. That's a sync buffer and has no effect here.

> I would say yes: we get 1-1-1-1-1 - 1-1-1-1-1-10min because of the done().

The buffer I was talking about: (the one after the candleBatcher emits a big candle, but before gekko runs lifecycle methods of strategy) the one inside the asyncIndicatorRunner. It's there because of the many issues regarding race conditions from async indicators (issues reported by users before that file existed).

> As long the updateOneMin has sync code there should be no problem. If you start async code in updateOneMin with callbacks than yes we might get into trouble. But now the update-method also expects only sync code.

This is not what I mean, we can easily document that users should not do this (the same as other lifecycle methods).

I am specifically talking about the about how many candles will ever pass through the stream (meaning hitting the processCandle of the tradingAdvisor without the previous one having called done). I agree that no writing stream should ever push more candles when the tradingAdvisor isn't done processing yet. But I am not sure if that's how all writing streams (called markets, 3 inside this repo and some more floating around the internet) are actually implemented. Or if they are not maybe the node streams API gives some guarantees?

Gekko had so many problems as a result of running async things related to strategies (especially in backtests, but also in realtime markets) that I only accept code for which we are sure of the behaviour.

Reply:
do you mean this 4 Readables? 

budfox (no get function) only feeds the db,
backtest (get function retrieves data from db),
leech (get function retrieves data from db),
importer (get function retrieves data from exchange and feeds db)

Only leech and backtest are used with strategies and they build very similar. I don't see any obstacles. 
In baseTradingMethod you call afterAsync() explicitly after asyncIndicatorRunner.processCandle() is finished, so that after propogateTick->update(candle) -> return done() line 127 is called. There is no way how  asyncIndicatorRunner can interfere with the gekko stream. 

Or do I miss something?

Reply:
> In baseTradingMethod you call afterAsync() explicitly after asyncIndicatorRunner.processCandle() is finished, so that after propogateTick->update(candle) -> return done() line 127 is called. 

True, what I mean is: when the baseTradingMethod will run asyncIndicators, and only after they are run the callback which came from the tradingAdvisor.processCandle. This is all great, however are we sure the stream will halt pumping new candles until this callback is ran? I am not sure the stream behaves like this in the realtime market. But I might be wrong, I just want to know for a fact given the amount of historical bugs regarding async candles.

Reply:
As far I understand, when we start realtime/backtesting we always start the leecher to read data from db: 

1. market data ( leecher readable Stream) pipes db data to 
2. gekkoStream (Writeable Stream) calls every plugin with processCandles (), and it will halt if done() function is not called.

if updateOneMin(candle) is sync and takes very long, than gekkoStream will wait and it could lead to : backpressure https://nodejs.org/en/docs/guides/backpressuring-in-streams/ 

"There are instances where a Readable stream might give data to the Writable much too quickly — much more than the consumer can handle!

When that occurs, the consumer will begin to queue all the chunks of data for later consumption. The write queue will get longer and longer, and because of this more data must be kept in memory until the entire process has completed."

There are solution to this problem which are described below in that article. 

The question is this the only concern?

Another question is what about async code in updateOneMin(candle) .  Because we might want to have other candles and we want to have asyncIndicators with this candles. Than we need to provide done callback like updateOneMin(candle, done)

Reply:
> Another question is what about async code in updateOneMin(candle)

The one minute candle works nice until you are going to use async code inside. I am also using this oneMin candles with async talib+tulip indicators, but in this case we need either a callback param or leave the current params and use the ECMAScript await/async/promise pattern.

Which flavor you take is a personal preference, but none of them gives you the ability to use the asyncIndicatorRunner out of the box with this custom/different candle size, or did I miss something?. How to tell this.addTulipIndicator that we want to use it with a different candle size? Therefor I added/suggested  async Tulip and Talib wrappers, see PR, to use these async indicators like any other native Gekko indicator (I am using it in my one minute candles side by side to other candle sizes)

Having this in mind it showed, that it will become a pretty nested callback flow in the end, so I personally went the await/async way. All I can say, it works pretty nice for me.

https://github.com/askmike/gekko/pull/2530/commits/affbb47de0db8b4f86d63abf7d21649d182fa65f

See line 102 for the one minute candles